src/sparse_blas/backends/cusparse/cusparse_error.hpp

```diff
@@ -0,0 +1,100 @@
+/***************************************************************************
+*  Copyright (C) Codeplay Software Limited
+*  Licensed under the Apache License, Version 2.0 (the "License");
+*  you may not use this file except in compliance with the License.
+*  You may obtain a copy of the License at
+*
+*      http://www.apache.org/licenses/LICENSE-2.0
+*
+*  For your convenience, a copy of the License has been included in this
+*  repository.
+*
+*  Unless required by applicable law or agreed to in writing, software
+*  distributed under the License is distributed on an "AS IS" BASIS,
+*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+*  See the License for the specific language governing permissions and
+*  limitations under the License.
+*
+**************************************************************************/
+
+#ifndef _ONEMKL_SPARSE_BLAS_BACKENDS_CUSPARSE_ERROR_HPP_
+#define _ONEMKL_SPARSE_BLAS_BACKENDS_CUSPARSE_ERROR_HPP_
+
+#include <string>
+
+#include <cuda.h>
+#include <cusparse.h>
+
+#include "oneapi/mkl/exceptions.hpp"
+
+namespace oneapi::mkl::sparse::cusparse {
+
+inline std::string cuda_result_to_str(CUresult result) {
+    switch (result) {
+#define ONEMKL_CUSPARSE_CASE(STATUS) \
+    case STATUS: return #STATUS
+        ONEMKL_CUSPARSE_CASE(CUDA_SUCCESS);
+        ONEMKL_CUSPARSE_CASE(CUDA_ERROR_NOT_PERMITTED);
+        ONEMKL_CUSPARSE_CASE(CUDA_ERROR_INVALID_CONTEXT);
+        ONEMKL_CUSPARSE_CASE(CUDA_ERROR_INVALID_DEVICE);
+        ONEMKL_CUSPARSE_CASE(CUDA_ERROR_INVALID_VALUE);
+        ONEMKL_CUSPARSE_CASE(CUDA_ERROR_OUT_OF_MEMORY);
+        ONEMKL_CUSPARSE_CASE(CUDA_ERROR_LAUNCH_OUT_OF_RESOURCES);
+        default: return "<unknown>";
+    }
+}
+
+#define CUDA_ERROR_FUNC(func, ...)                                                  \
+    do {                                                                            \
+        auto res = func(__VA_ARGS__);                                               \
+        if (res != CUDA_SUCCESS) {                                                  \
+            throw oneapi::mkl::exception("sparse_blas", #func,                      \
+                                         "cuda error: " + cuda_result_to_str(res)); \
+        }                                                                           \
+    } while (0)
+
+inline std::string cusparse_status_to_str(cusparseStatus_t status) {
+    switch (status) {
+#define ONEMKL_CUSPARSE_CASE(STATUS) \
+    case STATUS: return #STATUS
```

Comment:
Could you keep this in one line? and just above the `switch` statement? It is confusing and takes a few seconds to understand what exactly is going on here otherwise.

Reply:
Don't touch `clang-format`, of course. However, is it rearranging things even if you move the `#define` before the switch statement (possibly with a blank line between) and the `#undef` _after_ the switch statement?

Reply:
Yes, looks like it still break the line even outside of the switch with a new line.

Reply:
Okay, never mind then.